File: BuildArrayFromPermutation.cpp
// Approach 1: Simply Storing Values

class Solution {
public:
    vector<int> buildArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        
        for(int i=0;i<n;i++){
            ans[i] = nums[nums[i]];
        }
        return ans;
    }
};

// Appraoch 2: Constant Space[Optimised] 

class Solution {
public:
    vector<int> buildArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        
        for(int i=0;i<n;i++){
            ans[i]+= n*(nums[nums[i]]%n);
        }
        for(int i=0;i<n;i++){
            ans[i] = ans[i]/n;
        }
        return ans;
    }
};

File: BuildArrayFromPermutation.java
// 1920

// Approach 1: 

class Solution {
    public int[] buildArray(int[] nums) {
        int ans[]=new int[nums.length];
        for(int i=0;i<nums.length;i++)
        {
           ans[i]=nums[nums[i]];
        }
        return ans;
    }
}

//Approach 2: Optimised
class Solution {
  public int[] buildArray(int[] nums) {
    final int n = nums.length;

    for (int i = 0; i < n; ++i)
      nums[i] += n * (nums[nums[i]] % n);

    for (int i = 0; i < n; ++i)
      nums[i] /= n;

    return nums;
  }
}

